// cpbooster/dist/src/Types/Veredict.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Veredict = void 0;
/*
    cpbooster "Competitive Programming Booster"
    Copyright (C) 2020  Sergio G. Sanchez V.

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */
var Veredict;
(function (Veredict) {
    Veredict[Veredict["AC"] = 0] = "AC";
    Veredict[Veredict["WA"] = 1] = "WA";
    Veredict[Veredict["TLE"] = 2] = "TLE";
    Veredict[Veredict["RTE"] = 3] = "RTE";
    Veredict[Veredict["CE"] = 4] = "CE";
    Veredict[Veredict["AC_WHEN_TRIMMED"] = 5] = "AC_WHEN_TRIMMED";
    Veredict[Veredict["ERROR"] = 6] = "ERROR";
    Veredict[Veredict["UNDETERMINED"] = 7] = "UNDETERMINED";
})(Veredict = exports.Veredict || (exports.Veredict = {}));
